In [1]:
ls

Datasets/       Shallow_DNN.py  Untitled.ipynb  weights


In [39]:
import numpy as np
import pickle as p
import sys

In [35]:
layers_dims = [12288, 128, 128, 128, 128, 128, 64, 64, 32, 1] #  4-layer model

In [19]:
wts = p.load(open("./weights","rb"))

In [4]:
wts.keys()

dict_keys(['b2', 'b1', 'b6', 'W4', 'b7', 'W6', 'W9', 'b8', 'b9', 'W1', 'b3', 'W3', 'b5', 'W2', 'W8', 'W5', 'W7', 'b4'])

In [5]:
w1 = wts['W1']
b1 = wts['b1']

In [8]:
print( "w1: {}".format(w1.shape) )
print( "b1: {}".format(b1.shape) )

w1: (128, 12288)
b1: (128, 1)


# Convert pickle file to ascii

### Dimension analysis

In [44]:
layers_dims = [12288, 128, 128, 128, 128, 128, 64, 64, 32, 1]

i = 1
n = len(layers_dims)
prev = layers_dims[0]

while( i<n ):
    cur = layers_dims[i]
    print( "{} x {} : {}".format(prev, cur, prev*cur) )
    prev = cur
    i += 1

12288 x 128 : 1572864
128 x 128 : 16384
128 x 128 : 16384
128 x 128 : 16384
128 x 128 : 16384
128 x 64 : 8192
64 x 64 : 4096
64 x 32 : 2048
32 x 1 : 32


### Create the target file

In [42]:
def create_simplified_file(wts_file, out_file):
    print( "Loading file...")
    wts = p.load(open(wts_file,"rb"))
    print( "Done" )
    wts_np = np.array([]);
    keys = list(wts.keys())
    f = open( out_file, "w" )
    for indx in range( int(len(keys) / 2 ) ):
        indx += 1
        Wn = "W"+str(indx)
        Bn = "b"+str(indx)
        Wn_ = np.array(wts[Wn]).flatten()
        Bn_ = np.array(wts[Bn]).flatten()
        print( "{}. {}".format( Wn, Wn_.shape) )
        print( "{}. {}".format( Bn, Bn_.shape) )
        for w in Wn_:
            f.write( "{} ".format(w) )
        f.write("\n")
        for b in Bn_:
            f.write( "{} ".format(b) )
        f.write("\n")
        sys.stdout.flush()
    f.close()

In [45]:
create_simplified_file( wts_file= "./weights", out_file= "weights.lst" )

Loading file...
Done
W1. (1572864,)
b1. (128,)
W2. (16384,)
b2. (128,)
W3. (16384,)
b3. (128,)
W4. (16384,)
b4. (128,)
W5. (16384,)
b5. (128,)
W6. (8192,)
b6. (64,)
W7. (4096,)
b7. (64,)
W8. (2048,)
b8. (32,)
W9. (32,)
b9. (1,)


## Summary
<hr/>

NN Archi => `[ 12288, 128, 128, 128, 128, 128, 64, 64, 32, 1 ]`
> 12288 x 128 : 1572864 <br/>
> 128 x 128 : 16384 <br/>
> 128 x 128 : 16384 <br/>
> 128 x 128 : 16384 <br/>
> 128 x 128 : 16384 <br/>
> 128 x 64 : 8192 <br/>
> 64 x 64 : 4096 <br/>
> 64 x 32 : 2048 <br/>
> 32 x 1 : 32 <br/>
<br/>
<hr/>

Each line of `weights.lst` has
> W1. (1572864,) <br/>
> b1. (128,) <br/>
> W2. (16384,) <br/>
> b2. (128,) <br/>
> W3. (16384,) <br/>
> b3. (128,) <br/>
> W4. (16384,) <br/>
> b4. (128,) <br/>
> W5. (16384,) <br/>
> b5. (128,) <br/>
> W6. (8192,) <br/>
> b6. (64,) <br/>
> W7. (4096,) <br/>
> b7. (64,) <br/>
> W8. (2048,) <br/>
> b8. (32,) <br/>
> W9. (32,) <br/>
> b9. (1,) <br/>